x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plugin/CursorTests.java

```diff
@@ -99,7 +60,73 @@ public void testVersionHandling() {
         );
     }
 
+    private static final NamedWriteableRegistry WRITEABLE_REGISTRY = new NamedWriteableRegistry(Cursors.getNamedWriteables());
+
     public static Cursor decodeFromString(String base64) {
-        return Cursors.decodeFromStringWithZone(base64).v1();
+        return decodeFromStringWithZone(base64, WRITEABLE_REGISTRY).v1();
+    }
+
+    public void testAttachingStateToCursor() {
+        Cursor cursor = randomSearchHitCursor();
+        ZoneId zone = randomZone();
+        String encoded = encodeToString(cursor, zone);
+
+        FormatterState state = randomFormatterState();
+        String withState = attachState(encoded, state);
+
+        Tuple<Cursor, ZoneId> decoded = decodeFromStringWithZone(withState, WRITEABLE_REGISTRY);
+        assertEquals(cursor, decoded.v1());
+        assertEquals(zone, decoded.v2());
+        assertEquals(state, Cursors.decodeState(withState));
+    }
+
+    public void testAttachingEmptyStateToCursor() {
+        Cursor cursor = randomSearchHitCursor();
+        ZoneId zone = randomZone();
+        String encoded = encodeToString(cursor, zone);
+
+        String withState = attachState(encoded, null);
+
+        Tuple<Cursor, ZoneId> decoded = decodeFromStringWithZone(withState, WRITEABLE_REGISTRY);
+        assertEquals(cursor, decoded.v1());
+        assertEquals(zone, decoded.v2());
+        assertNull(Cursors.decodeState(withState));
+    }
+
+    public void testAttachingStateToEmptyCursor() {
+        Cursor cursor = Cursor.EMPTY;
+        ZoneId zone = randomZone();
+        String encoded = encodeToString(cursor, zone);
+
+        FormatterState state = randomFormatterState();
+        String withState = attachState(encoded, state);
+
+        assertEquals(StringUtils.EMPTY, withState);
+
+        Tuple<Cursor, ZoneId> decoded = decodeFromStringWithZone(withState, WRITEABLE_REGISTRY);
+        assertEquals(cursor, decoded.v1());
+        assertNull(decoded.v2());
+        assertNull(Cursors.decodeState(withState));
+    }
+
+    public void testAttachingStateToCursorFromOtherVersion() {
+        Cursor cursor = randomSearchHitCursor();
+        ZoneId zone = randomZone();
+        String encoded = encodeToString(cursor, Version.V_7_17_1, zone);
```

Comment:
Is there anything in particular with Version.V_7_17_1 or there can be any version?

Reply:
nope, I think it can be any version < current. But we do not have a `randomVersion` yet.

Reply:
I think we have a variant of randomVersion: `VersionUtils.randomVersionBetween`

Reply:
Turns out I didn't search hard enough: There is `VersionUtils.randomVersion`

Reply:
argh, GH should have live reload for comments ;)